src/log_batch.rs

```diff
@@ -485,115 +472,128 @@ where
         for item in self.items.iter_mut() {
-            item.encode_to(&mut vec, &mut entries_size).unwrap();
+            item.encode_entries_to(&mut buf, &mut 0)?;
+        }
+        let compression_type =
+            if compression_threshold > 0 && (buf.len() - HEADER_LEN) > compression_threshold {
+                buf = lz4::encode_block(&buf[HEADER_LEN..], HEADER_LEN, 4);
+                CompressionType::Lz4
+            } else {
+                CompressionType::None
+            };
+        let checksum = crc32(&buf[HEADER_LEN..]);
+        buf.encode_u32_le(checksum)?;
+
+        let offset = buf.len() as u64;
+
+        // encode footer (entries_index / commands / kvs)
+        let count = self.items.len() as u64;
+        buf.encode_var_u64(count)?;
+        for item in self.items.iter() {
+            item.encode_to(&mut buf)?;
         }
+        let checksum = crc32(&buf[offset as usize..]);
+        buf.encode_u32_le(checksum)?;
 
-        let compression_type = if compression_threshold > 0 && vec.len() > compression_threshold {
-            vec = lz4::encode_block(&vec[HEADER_LEN..], HEADER_LEN, 4);
-            CompressionType::Lz4
-        } else {
-            CompressionType::None
-        };
+        // write header
+        let len = ((buf.len() as u64) << 8) | u64::from(compression_type.to_byte());
 
-        let checksum = crc32(&vec[8..]);
-        vec.encode_u32_le(checksum).unwrap();
-        let len = vec.len() as u64 - 8;
-        let mut header = len << 8;
-        header |= u64::from(compression_type.to_byte());
-        vec.as_mut_slice().write_u64::<BigEndian>(header).unwrap();
+        (&mut buf[..HEADER_LEN]).write_u64::<BigEndian>(len)?;
+        (&mut buf[8..HEADER_LEN]).write_u64::<BigEndian>(offset)?;
 
-        let batch_len = (vec.len() - 8) as u64;
+        // update entries_index
         for item in self.items.iter_mut() {
             if let LogItemContent::Entries(entries) = &mut item.content {
-                entries.update_compression_type(compression_type, batch_len as u64);
+                for ei in entries.entries_index.iter_mut() {
+                    ei.compression_type = compression_type;
+                    ei.entries_offset = HEADER_LEN as u64; // need to add batch offset of the file after written.
+                    ei.entries_len = offset as usize - HEADER_LEN;
+                }
             }
         }
-
-        Some(vec)
+        Ok(Some(buf))
     }
 
     // Don't account for compression and varint encoding.
     pub fn approximate_size(&self) -> usize {
         if self.items.is_empty() {
             0
         } else {
-            8 /*len*/ + 8 /*count*/ + self.items_size + 4 /*checksum*/
+            8 /*len*/ + 8 /*section offset*/ + 8/*items count*/
+            +self.items_approximate_size + CHECKSUM_LEN * 2 /*checksum*/
         }
     }
+
+    /// min buffer size for recover memtable ()
+    pub fn recovery_size(buf: &mut SliceReader<'_>) -> Result<usize> {
```

Comment:
Not used?

Reply:
It's from the first version of the new log foramt. I forgot to remove them.